cmd/client_test.go

```diff
@@ -0,0 +1,224 @@
+func setupHermes(hermes *hermes.Hermes) {
+	err := hermes.InitialiseCloudproberFromConfig(cpCfg)
+	if err != nil {
+		glog.Fatalf("Cloudprober could not be initialised, err: %v", err)
+	}
+}
+
+// TestMain is used to setup Cloudprober before tests are run.
+// This is necessary for there to be a gRPC server to make RPCs to.
+func TestMain(m *testing.M) {
+	test_hermes := &hermes.Hermes{}
+
+	setupHermes(test_hermes)              // Set up Hermes and initialise Cloudprober
+	defer test_hermes.CancelCloudprober() // Cancel Cloudprober after tests are completed
+
+	// Sets up web UI for cloudprober.
+	web.Init()
+
+	// Start running Cloudprober instance from Hermes context
+	cloudprober.Start(test_hermes.Ctx)
+
+	code := m.Run()
+	os.Exit(code)
+}
+
+// generationRedisProbeDef generates an inline probe definition of the RedisProbe probe extension.
+// The RedisProbe extension is supplied as an example extension in Cloudprober.
+func generateRedisProbeDef(name string) *probes_configpb.ProbeDef {
+	probeDef := &probes_configpb.ProbeDef{ // Create probe def
+		Name: proto.String(name),
+		Type: probes_configpb.ProbeDef_EXTENSION.Enum(),
+		Targets: &targetspb.TargetsDef{
+			Type: &targetspb.TargetsDef_DummyTargets{},
+		},
+	}
+
+	// Add RedisProbe extension to probeDef
+	op := myprobe.ProbeConf_Op.Enum(myprobe.ProbeConf_SET)
+	proto.SetExtension(probeDef, myprobe.E_RedisProbe, &myprobe.ProbeConf{Op: op, Key: proto.String("testkey"), Value: proto.String("testval")})
+	return probeDef
+}
+
+// TestRegisterAndAddProbe tests that an extension probe type can be registered and added to Cloudprober without error.
+func TestRegisterAndAddProbe(t *testing.T) {
+	cpc := &CloudproberClient{}
+	cpc.InitClient(rpcServer)
+
+	// Create a probe and then register and add it to the prober
+	err := cpc.RegisterAndAddProbe(int(myprobe.E_RedisProbe.Field), generateRedisProbeDef(testProbe0), &myprobe.Probe{})
+	if err != nil {
+		t.Errorf("Probe not correctly registered and added to Cloudprober, error: %v", err)
+	}
+
+	// Check if the probe was added correctly
+	resp, errResp := cpc.ListProbes()
+	if errResp != nil {
+		t.Errorf("ListProbes() failed, error: %v", errResp)
+	}
+
+	respProbes := resp.GetProbe()
+
+	if respProbes[0].GetName() != testProbe0 {
+		t.Errorf("Expected probe %s to be running, got probe %s running", testProbe0, respProbes[0].GetName())
+	}
+
+	// Try adding a second probe to see if the registration is still in place
+	err = cpc.RegisterAndAddProbe(int(myprobe.E_RedisProbe.Field), generateRedisProbeDef(testProbe1), &myprobe.Probe{})
+
+	// Check if the probe was added correctly
+	resp, errResp = cpc.ListProbes()
+	if errResp != nil {
+		t.Errorf("ListProbes() failed, error: %v", errResp)
+	}
+
+	respProbes = resp.GetProbe()
+
+	if respProbes[1].GetName() != testProbe1 {
+		t.Errorf("Expected probe %s to be running, got probe %s running", testProbe0, respProbes[1].GetName())
```

Comment:
From the if statement in the line above, I thought that the expected probe you wanted was testProbe1 instead of testProbe0. Please make these consistent.

Reply:
I have refactored this into a table driven test, and removed this inconsistency.